### DISCUSSION TITLE
Bypassing vcpkg for a specific package or dependency only

### DISCUSSION THREAD

**Discussion by craigscott-crascit** [microsoft/vcpkg#22836] [Q&A] [answered]
I have a use case where I would like to prevent vcpkg from providing a specific package, and instead let `find_package()` pick up whatever it would have found had vcpkg not been involved. As a specific example, let's say I need to pick up the python version already set up in the build environment (for reasons). I specifically want to avoid vcpkg trying to use the python3 dependency it would otherwise provide.

Things I've explored but which don't quite fit:

* Using `VCPKG_PREFER_SYSTEM_LIBS`. This is global and I need something that can be targeted at specific packages. It also doesn't prevent vcpkg from taking control of version requirements. Adding an `overrides` entry to the `vcpkg.json` manifest would not solve that, because that file would need to be stored in the project repo, but the actual python version in the environment would be chosen by the user at configuration time.
* Creating my own registry with the package details I intend to use. Seems overly complicated and I'd have to go introspecting details of the python used, then dynamically create a registry for vcpkg to use. Seems far to complicated for what I'm trying to achieve.
* Create an overlay. This would solve the versioning issue, but I'm struggling to find a good example of how to easily define an overlay that simply picks up whatever would be found by CMake if vcpkg wasn't involved. In the python example, I'd just want `find_package(Python3)` to find the python available on the PATH.

I understand the dangers of doing what I'm trying to do. I also understand it goes against the thinking behind vcpkg. Nevertheless, I'm interested in whether there's a relatively simple way to achieve this.

**Comment by Neumann-A** [comment, marked as answer]
> Create an overlay. This would solve the versioning issue, but I'm struggling to find a good example of how to easily define an overlay that simply picks up whatever would be found by CMake if vcpkg wasn't involved.

Create an empty port in the overlay for whatever you want the system to provide.  The only thing the `portfile.cmake` needs is: `set(VCPKG_POLICY_EMPTY_PACKAGE enabled)`

**Reply by craigscott-crascit** [reply]
I still need `find_package()` to find the package though, I only want it to avoid looking in vcpkg's ports, not skip it entirely. Would your suggestion achieve that?

**Reply by Neumann-A** [reply]
> I still need find_package() to find the package though

vcpkg does not block normal `find_package()` behavior. 

>  I only want it to avoid looking in vcpkg's ports, not skip it entirely.

The only way to achieve this is it to make `find_library/find_package` not find the files in vcpkg which it normally searches for. Since vcpkg adds itself to the `CMAKE_PREFIX_PATH` the only way to not find stuff within vcpkg is to remove the files completely or explicitly feed cmake with the result of the search variables. In the case of python this would probably be `Python3_ROOT_DIR` ?
But I would say it is way safer to just make a port empty and let cmake figure out the rest.

**Reply by Neumann-A** [reply]
>  result of the search variables. 

In the python3 case this includes what `vcpkg-cmake-wrapper.cmake` is looking for. (so making the port empty is way easier.)

**Reply by craigscott-crascit** [reply]
Thanks. An empty port does then look close to what I need. Looks like I still have to interrogate the python executable to extract its version and write the vcpkg.json file in the overlay, which is a bit of a bummer, but I guess not a deal breaker. I'll tinker with this a bit more and see if it plays out for my use case.

**Reply by craigscott-crascit** [reply]
Confirmed this approach works for my use case. Thanks for the insights.
